### PR TITLE
DOC-#57585: Add `Use Modin` section on `Scaling to large datasets` page

### LINKED ISSUE #57585: DOC: Add `Use Modin` section on `Scaling to large datasets` page
### Pandas version checks

- [X] I have checked that the issue still exists on the latest versions of the docs on `main` [here](https://pandas.pydata.org/docs/dev/)

### Location of the documentation

https://pandas.pydata.org/docs/user_guide/scale.html

### Documentation problem

While there are different libraries to scale large datasets, the only option shown on `scale.rst` page is Dask. It could be useful and valuable for users to get more options out of the page. One of the options could be Modin, which has a drop-in replacement API for pandas. It would be nice to show how Modin can speed up pandas operations, as well as to be used along with pandas in a single flow to get ultimate performance and reduce users's waiting time.

### Suggested fix for documentation

Add a section about Modin.

### PR DESCRIPTION
**Pull request by YarShev**
- [x] closes #57585  (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: doc/source/user_guide/scale.rst
Comment (rhshadrach) [inline doc/source/user_guide/scale.rst:413]: I'm guessing you mean for this to be a static code block - use `.. code-block:: ipython` instead. Otherwise this will be run when building the docs, leaving behind a large file and adding a significant amount of build time.

In addition, I think comparing to an ill-performant pandas version is misleading. A proper comparison here is to`pandas_df = pandas_df + 0.01`

File: doc/source/user_guide/scale.rst
Comment (rhshadrach) [inline doc/source/user_guide/scale.rst:426]: `rather than a number of small ones` I think?

File: doc/source/user_guide/scale.rst
Comment (rhshadrach) [inline doc/source/user_guide/scale.rst:462]: This looks like it could be misleading, but is in fact not. It seems to me a reader might think that the same change to the pandas code would also decrease its execution time but pandas remains roughly the same. I think it'd be clearer to add timing for the pandas equivalent here too:

```
pandas_subset = pandas_df.iloc[:100000]
pandas_subset = pandas_subset / pandas_subset.sum()
```

File: doc/source/user_guide/scale.rst
Comment (rhshadrach) [inline doc/source/user_guide/scale.rst:382]: nit: This is a run on sentence. Maybe just split off the last part?

> Modin_ is a scalable dataframe library, which has a drop-in replacement API for pandas and provides the ability to scale pandas workflows across nodes and CPUs available. It is also able to work with larger than memory datasets.

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Thanks for the PR!

File: doc/source/user_guide/scale.rst
Comment (YarShev) [reply doc/source/user_guide/scale.rst:382]: Done, thanks for the suggestion.

File: doc/source/user_guide/scale.rst
Comment (YarShev) [reply doc/source/user_guide/scale.rst:413]: > I'm guessing you mean for this to be a static code block - use `.. code-block:: ipython` instead. Otherwise this will be run when building the docs, leaving behind a large file and adding a significant amount of build time.

Applied `.. code-block:: ipython`. Thanks.

> In addition, I think comparing to an ill-performant pandas version is misleading. A proper comparison here is to `pandas_df = pandas_df + 0.01`

I understand that the binary operation is much more efficient than appliying a function with `map`. Since the user may apply any function on the data, I wanted to show how it performs with pandas and Modin. Do you mind we leave this as is?

File: doc/source/user_guide/scale.rst
Comment (YarShev) [reply doc/source/user_guide/scale.rst:426]: Good catch! I meant `rather a big number of small ones`. Fixed.

File: doc/source/user_guide/scale.rst
Comment (YarShev) [reply doc/source/user_guide/scale.rst:462]: Added, thanks!

**Comment by YarShev** [comment]
@rhshadrach, could you shed some light on why `Doc Build` is failing and what we should do to fix it?

File: doc/source/user_guide/scale.rst
Comment (jbrockmendel) [inline doc/source/user_guide/scale.rst:380]: "has a drop-in" -> "aims to be a drop-in" would be more accurate

Comment (YarShev) [reply doc/source/user_guide/scale.rst:380]: @jbrockmendel, could you please elaborate on what gave you this opinion? The user can proceed using pandas API after replacing the import statement.

File: doc/source/user_guide/scale.rst
Comment (rhshadrach) [reply doc/source/user_guide/scale.rst:413]: > Do you mind we leave this as is?

It seems to me that a reader could easily come away from the existing example thinking "modin can speed up simple arithmetic operations by 25x over pandas" - a statement which is not true. In fact, I'm seeing 0.26 seconds when doing `pandas_df + 0.01` on my machine. Because of this - I do mind.

I see a few ways forward. I think you are trying to convey the benefit of using arbitrary Python functions with modin - this could be stated up front explicitly and a note at the end benchmarking the more performant pandas version. Alternatively, you could use a Python function for which there is no simple pandas equivalent.

File: doc/source/user_guide/scale.rst
Comment (rhshadrach) [reply doc/source/user_guide/scale.rst]: nit: "big number" sounded off to me for some reason. A couple of quick searches I'm seeing very consistent recommendations to use "large number" instead, e.g. [Cambridge Dictionary](https://dictionaryblog.cambridge.org/2016/03/16/a-whole-bunch-of-stuff-talking-about-quantities-and-numbers-1/).

File: doc/source/user_guide/scale.rst
Comment (rhshadrach) [reply doc/source/user_guide/scale.rst]: I'm not seeing the same thing on my machine.

L438: You report 293ms, I get 52ms
L462: You report 72.2ms, I get 69ms

Can you check your timings again? If they hold up, what version of pandas/NumPy are you running?

**Comment by rhshadrach** [comment]
> @rhshadrach, could you shed some light on why `Doc Build` is failing and what we should do to fix it?

L474, documentation is spelled incorrectly.

File: doc/source/user_guide/scale.rst
Comment (jbrockmendel) [reply doc/source/user_guide/scale.rst:380]: In my experience working on modin lots of behaviors didn't quite match.  Recall I presented a ModinExtensionArray and about half of the tests could not be made to pass because of these.

Comment (YarShev) [reply doc/source/user_guide/scale.rst:380]: I see, thanks! Yes, "aims to be a drop-in" would be more accurate. Fixed.

File: doc/source/user_guide/scale.rst
Comment (YarShev) [reply doc/source/user_guide/scale.rst]: Fixed, thanks!

File: doc/source/user_guide/scale.rst
Comment (YarShev) [reply doc/source/user_guide/scale.rst:413]: Grabbed an example from https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.map.html to use a custom function for applying on the data. Hope it is okay for now.

File: doc/source/user_guide/scale.rst
Comment (YarShev) [reply doc/source/user_guide/scale.rst]: The time might be different for my and your runs. I have Intel(R) Xeon(R) Platinum 8468. Reran again and updated the timings. I have these pandas and numpy versions installed.

```bash
pandas                    2.2.0
numpy                     1.26.0
```

**Comment by YarShev** [comment]
> > @rhshadrach, could you shed some light on why `Doc Build` is failing and what we should do to fix it?
> 
> L474, documentation is spelled incorrectly.

Oh, I see, thanks! FIxed.

**Comment by datapythonista** [comment]
I see the Dask section was added 5 years ago, where there were probably no other alternatives. I think it's surely valuable for users to know about other tools to distribute work with a pandas API, but I don't think our user guide is the place to have and have to maintain a short tutorial of Dask. Modin, PySpark/Koalas and any other system that provides a distributed pandas-like API.

What do you think if you have in each project documentation a page for pandas users that we can simply link here? @phofl @YarShev @HyukjinKwon

**Comment by YarShev** [comment]
> I see the Dask section was added 5 years ago, where there were probably no other alternatives. I think it's surely valuable for users to know about other tools to distribute work with a pandas API, but I don't think our user guide is the place to have and have to maintain a short tutorial of Dask. Modin, PySpark/Koalas and any other system that provides a distributed pandas-like API.
> 
> What do you think if you have in each project documentation a page for pandas users that we can simply link here? @phofl @YarShev @HyukjinKwon

While other projects can have a documentation page about their tool and pandas as to which library to use and when, I think mentioning other alternatives to pandas in pandas documentation itself is valuable to users since pandas is much more popular than the alternatives, and many more people will be able to see concrete options for efficient data processing.

**Comment by datapythonista** [comment]
I don't disagree with you, and I think it's valuable for users to know the other packages in the ecosystem to scale what pandas can't handle.

My point is that in my opinion the pandas codebase is not where so much information from Dask or Modin should exist. I think we should have a paragraph so users know about the tools and what to expect about them, and all the details could be in a landing page for pandas users that you maintain in your docs, keep updated, and that we simply have a link here next to the summary of what each tool does.

Just my personal opinion, but this should be as good for the users, but easier to maintain and keep updated.

**Comment by YarShev** [comment]
I tried to make the changes as concise as possible giving some pros and cons of Modin and when it should be used. I don't see anything in the changes that could be hard to maintain in future. If you have any improvements we could introduce in the paragraph, please let me know. It would be also great to hear opinions from other reviewers. cc @rhshadrach, @jbrockmendel

**Comment by rhshadrach** [comment]
Building on what @datapythonista said - I don't see much advantage in having the documentation as it is in this PR vs just the intro paragraph ending with a line like `See more information in [modin's documentation](https://modin.readthedocs.io/en/stable/getting_started/quickstart.html#example-instant-scalability-with-no-extra-effort).` (or a better landing spot if there is one).

> I don't see anything in the changes that could be hard to maintain in future.

Changes to any of the third party packages currently listed would necessitate a change to the pandas documentation and is therefore quite likely to go stale. In addition each new tool that comes along we have to review another PR. It is all around less maintenance for us to have a summary paragraph and a link to the projects documentation where they compare to pandas.

What is the disadvantage by having a link rather than the documentation live in pandas itself?

**Comment by HyukjinKwon** [comment]
@datapythonista what about pandas here points out the tutorial of individual project?

Basically I support:

> he intro paragraph ending with a line like See more information in [modin's documentation](https://modin.readthedocs.io/en/stable/getting_started/quickstart.html#example-instant-scalability-with-no-extra-effort). (or a better landing spot if there is one).

FWIW, Koalas has been migrated to PySpark as pandas API on Spark component so I have to update it in any event.

**Comment by datapythonista** [comment]
> @datapythonista what about pandas here points out the tutorial of individual project?

Sounds great. My idea is that users reaching our page about scaling should know about the relevant options, but without pandas having lots of details in our docs when we can simply link to each projects docs. The exact page you link to I don't know, maybe a dedicated landing page for users coming from pandas, maybe the home of the Koalas documentation, no strong preference from my side.

Also worth mentioning that we already do it in the Ecosystem page in our website. Maybe for now I can just copy what it's there in a separate PR, and you all can review it or follow up with any change if needed. What do you think?

File: doc/source/user_guide/scale.rst
Comment (datapythonista) [inline doc/source/user_guide/scale.rst:381]: Small thing, but maybe you can add the part about execution engines here, since in the next paragraph you say `Modin distributes computation across nodes and CPUs available` which seems to pretty much repeat the same thing again to conextualize that part?

File: doc/source/user_guide/scale.rst
Comment (datapythonista) [inline doc/source/user_guide/scale.rst:395]: Not sure if it's just me, but the part about partitioning seems too technical and out of context for this short intro. Feel free to disagree if you think it's key for users to know this asap when considering Modin.

File: doc/source/user_guide/scale.rst
Comment (datapythonista) [inline doc/source/user_guide/scale.rst:398]: Maybe I'm being picky, but since this is in the pandas docs, I'd remove the last part that sounds like the end of a sales pitch. ;) 

In my opinion, just: "For more information refer to `Modin's documentation`_ or the `Modin's tutorials`_. " would feel more appropriate.

**Review by datapythonista — COMMENTED** [review]
Thanks for the flexibility and for updating this @YarShev. I added couple of very opinionated comments that in my opinion would make this a bit more concise and to the point. Feel free to disagree if you don't like my comments.

File: doc/source/user_guide/scale.rst
Comment (rhshadrach) [reply doc/source/user_guide/scale.rst:398]: +1

Comment (YarShev) [reply doc/source/user_guide/scale.rst:398]: Agree with you. Changed.

File: doc/source/user_guide/scale.rst
Comment (YarShev) [reply doc/source/user_guide/scale.rst:395]: Since this page is not a getting started, I think it is fine to say some words here about partitioning, which is flexible along columns and row.

File: doc/source/user_guide/scale.rst
Comment (YarShev) [reply doc/source/user_guide/scale.rst:381]: I wasn't able to rephrase the first and the second paragraphs by following your comment. To me the current state of paragraphs is pretty concise.

**Comment by YarShev** [comment]
Ready for the next round of review. Hope it is fine now 😄

**Review by rhshadrach — APPROVED** [review]
Thanks for the changes @YarShev - this looks good to me pending the current discussions.

File: doc/source/user_guide/scale.rst
Comment (rhshadrach) [reply doc/source/user_guide/scale.rst:381]: To me, this intro paragraph aims to be a high level description of Modin. With that in mind, getting into execution engines here seems undesirable.

File: doc/source/user_guide/scale.rst
Comment (rhshadrach) [reply doc/source/user_guide/scale.rst:395]: I think being both columnar and row-wise is an important feature to mention (since some things only scale well in one axis).

**Comment by YarShev** [comment]
@datapythonista, @HyukjinKwon, just a friendly reminder for review.

File: doc/source/user_guide/scale.rst
Comment (HyukjinKwon) [inline doc/source/user_guide/scale.rst:377]: Just my two cents. I would just simply remove both Dask and Modin here, add a link to https://github.com/pandas-dev/pandas/blob/a0784d2c59c04e0a9e1ef22d696adb5768b5541e/web/pandas/community/ecosystem.md?plain=1#L344

A lot of information here is duplicated with Dask vs Modin and others too such as Pandas API on Spark (Koalas). They all aim drop-in replacement, they all out-of-core in multiple nodes, partitioning, scalability.

Comment (HyukjinKwon) [reply doc/source/user_guide/scale.rst:377]: Or, I would at least deduplicate them here if we want to keep those sections. e.g.,

```
# Using thirdparty libraries

multiple nodes blah blah..

## Dask

oneliner explanation

Link

## Modin

oneliner explanation

Link

## PySpark: Pandas API on Spark

oneliner explanation

Link


```

Comment (YarShev) [reply doc/source/user_guide/scale.rst:377]: > A lot of information here is duplicated with Dask vs Modin and others too such as Pandas API on Spark (Koalas). They all aim drop-in replacement, they all out-of-core in multiple nodes, partitioning, scalability.

I would argue against the statement `They all aim drop-in replacement`. Modin is supposed to work after changing the import statement `import pandas as pd` to `import modin.pandas as pd` and vice versa, while Dask or Koalas require the code to be rewritten to the best of my knowledge. Anyway, if everyone thinks that `Scaling to large datasets` page should only contain pandas related stuff, I can go ahead and remove Modin and Dask related stuff from that page and move the changes from this PR to the ecosystem page. Thoughts?

Comment (HyukjinKwon) [reply doc/source/user_guide/scale.rst:377]: I'm not really a maintainer here and don't have a super strong opinion. I'll defer to pandas maintainers.

**Review by datapythonista — APPROVED** [review]
I agree with @HyukjinKwon about the one-liner, as I mentioned earlier I find the way things are phrased now too verbose and a bit repetitive considering we want something concise, but no big deal, I'm ok to get this merged as is, and we can iterate further, once we reduce the Dask section and add PySpark/Koalas.

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @YarShev. Agreed that eventually this should just link to the ecosystem docs, but this is good as is.

**Pull request merged**
